### ISSUE TITLE
Develop example of tooltip design pattern

### ISSUE THREAD

**Issue by mcking65** [w3c/aria-practices#127]
The tooltip design pattern is at:
http://w3c.github.io/aria-practices/#tooltip

**Comment by Veyfeyken** [comment]
Any chance this will go live in the near future? I'm looking for an accessible example to point to.

**Comment by Malvoz** [comment]
@Veyfeyken 

While you are waiting for the w3c developers to design a tooltip pattern, you might want to have a look at [Heydon Pickering](https://github.com/Heydon)'s [**tooltip component**](https://inclusive-components.design/tooltips-toggletips/).

**Comment by MichielBijl** [comment]
Work on this is being done [on CodePen](https://codepen.io/Moiety/pen/LaPvWy). You can also view the [debug page](https://s.codepen.io/Moiety/debug/LaPvWy) which is a lot easier to navigate with AT.

Current open issues with the implementation:

- [ ] Doesn’t detect window bounds
- [ ] Doesn’t conform to [WCAG 2.1 SC 1.4.13 Content on Hover or Focus](https://www.w3.org/TR/WCAG21/#content-on-hover-or-focus)

Any help is greatly appreciated.

**Comment by krixian** [comment]
@MichielBijl It's looking good! Is there any description on how overflowing tooltips should behave?

**Comment by weboverhauls** [comment]
Yes, looking good! Would also be great to have a tooltip example on something other than a button such as a link/anchor, text input, and possibly regular text.

**Comment by mfairchild365** [comment]
@MichielBijl that is a great example. One question immediately comes to mind: the example uses a `<button>` element to trigger the tooltip. Because the tooltip is automatically displayed when the button recieves both hover and focus events, activating the button element doesn't actually do anything. My hunch is that users might find a button that doesn't do anything a bit confusing and unexpected.

As indicated in #128 the presence of a button might indicate a 'dialog toolip' pattern rather than a traditional tooltip. In any case, I think it might be worth while to think about what the role of the of the element that triggers the tooltip should be. My gut says that no role would be sufficent, but perhaps I'm way off base.

**Comment by MichielBijl** [comment]
> @MichielBijl It's looking good! Is there any description on how overflowing tooltips should behave?

Do you mean when there’s more text than fits ion the tooltip?

> Yes, looking good! Would also be great to have a tooltip example on something other than a button such as a link/anchor, text input, and possibly regular text.

That’s a good idea! I will add that.

> @MichielBijl that is a great example. One question immediately comes to mind: the example uses a `<button>` element to trigger the tooltip. Because the tooltip is automatically displayed when the button receives both hover and focus events, activating the button element doesn't actually do anything. My hunch is that users might find a button that doesn't do anything a bit confusing and unexpected.

Hmm, the buttons in the example would do something if they were implemented on a page. The heart/like button is what you’d find in Twitter’s UI for example. I can trigger some kind of alert to show that it did something. Would that clarify it?

> As indicated in #128 the presence of a button might indicate a 'dialog toolip' pattern rather than a traditional tooltip.
> 
> In any case, I think it might be worthwhile to think about what the role of the of the element that triggers the tooltip should be (when the tooltip is not attached to a form control or another element that performs a separate action, and the tooltip content is shown on touch, hover, and focus events). My gut says that no role would be sufficient, but perhaps I'm way off base.

While tooltips can certainly be used for other things; their main usage would still be to clarify buttons/other controls no?

From @heydon’s article:
> Literally "tips for tools", they are little bubbles of information that clarify the purpose of otherwise ambiguous controls/tools.

As for the controls not triggering anything, I can look into that. I think if I add an action to them the tooltips won’t show anymore on touch devices (which also shows how useless tooltips are for UX patterns).

**Comment by mfairchild365** [comment]
> Hmm, the buttons in the example would do something if they were implemented on a page. The heart/like button is what you’d find in Twitter’s UI for example. I can trigger some kind of alert to show that it did something. Would that clarify it?

Good point, I think that would clarify it.

**Comment by carmacleod** [comment]
Great examples, @ZoeBijl !

I think you've covered all of the WCAG 2.1 SC 1.4.13 bases, because they are Dismissable, Hoverable and Persistent. Did you have something specific in mind when you wrote "Doesn’t conform to WCAG 2.1 SC 1.4.13 Content on Hover or Focus" in https://github.com/w3c/aria-practices/issues/127#issuecomment-467152945 ?

**Comment by carmacleod** [comment]
I am pretty sure that @ZoeBijl's examples pass [WCAG 2.1 SC 1.4.13](https://www.w3.org/TR/WCAG21/#content-on-hover-or-focus).
@jnurthen Can you please confirm for me?

Here's the text of the SC:
> Where receiving and then removing pointer hover or keyboard focus triggers additional content to become visible and then hidden, the following are true:
>
> - **Dismissable**
> A mechanism is available to dismiss the additional content without moving pointer hover or keyboard focus, unless the additional content communicates an input error or does not obscure or replace other content;
> - **Hoverable**
> If pointer hover can trigger the additional content, then the pointer can be moved over the additional content without the additional content disappearing;
> - **Persistent**
> The additional content remains visible until the hover or focus trigger is removed, the user dismisses it, or its information is no longer valid.

The example is here: https://s.codepen.io/Moiety/debug/LaPvWy
Here's my reasoning:
- Dismissable
  - hover over one of the buttons, tooltip appears. Type Escape, tooltip is dismissed.
  - tab to one of the buttons, tooltip appears. Type Escape, tooltip is dismissed.
- Hoverable
  - hover over one of the buttons, tooltip appears. Move mouse over tooltip, tooltip stays.
  - (tooltip does not timeout at all, i.e. if you hover over button and stay there, tooltip remains visible)
- Persistent
  - hover over one of the buttons, tooltip appears. Move mouse cursor out of button (and out of tooltip), tooltip is dismissed.
  - tab to one of the buttons, tooltip appears. Type tab again, tooltip is dismissed.

**Comment by carmacleod** [comment]
@ZoeBijl 
Note that `aria-hidden` defaults to "undefined", so please use `aria-hidden="true"`.

The other issue mentioned in yesterday's call is that the tooltip is shown immediately without any delay. May want to implement a "hover intent" pattern to ensure that a mouse user actually intended to hover.
For example, hover over github's toolbar buttons above (Bold, Italic, etc). If you move past the button, the tooltip will not display, but if you go slowly or stop on a button, then the tooltip displays.

**Comment by ZoeBijl** [comment]
That’s a good suggestion. I’ll move the code I have so far to the APG repo so we can collaborate on it more easily. My main issue is bounding box detection and I would appreciate help with that.

**Comment by maschad96** [comment]
@ZoeBijl is this something I can contribute to? I'd be happy to help if there's room for that.

**Comment by ZoeBijl** [comment]
> @ZoeBijl is this something I can contribute to? I'd be happy to help if there's room for that.

I’d love some help with the bounding box stuff. I’m back on Monday and will ping you with a location of this stuff :)

**Comment by JAWS-test** [comment]
When the pattern is finished, it may also be used for WCAG. See the discussion:
* https://github.com/w3c/wcag/issues/915
* https://github.com/w3c/wcag/issues/914

**Comment by LaurenceRLewis** [comment]
I noticed that the tooltips when viewed on iPhone and open when touched, one tooltip will close when the other is opened. However, the tool tips do not close when touched again (opened) or when touching on the surrounding background. From a usability perspective how is a touch user meant to close an open tooltip?

Thanks

https://s.codepen.io/Moiety/debug/LaPvWy

**Comment by anevins12** [comment]
I may be misunderstanding @LaurenceRLewis , but does Scott's examples hinder someone using a screen magnifier that could be panning over left, right, down or up around the tooltip to see it all? By panning the tooltip disappears.

**Comment by LaurenceRLewis** [comment]
@anevins12 I admit I did not consider panning outside the tooltip area. I have removed the link until I run more testing.

**Comment by anevins12** [comment]
@LaurenceRLewis I think Scott's examples are a great example and should totally be showcased. I think that I had a slight misunderstanding in the WCAG On Hover or Focus criterion in that I misread this bit:
>  then the pointer can be moved over the additional content without the additional content disappearing;

It doesn't say the pointer can be moved around the additional content. The criterion is specific to moving the hover over the additional content.

Now that I think of it clearer, we see that behaviour all the time in dropdown navigation menus. With the pattern so common I think it would be fine to have Scott's examples recommended here.

Sorry for the confusion!

**Comment by LaurenceRLewis** [comment]
@anevins12 no worries, I ask questions often because I misinterpret a Success Criterion.

**Comment by idoros** [comment]
> - Persistent
>     - hover over one of the buttons, tooltip appears. Move mouse cursor out of button (and out of tooltip), tooltip is dismissed.
>     - tab to one of the buttons, tooltip appears. Type tab again, tooltip is dismissed.

@carmacleod, I'm sorry if I missed something, but what should be the cross behavior between focus and hover.

**case 1:** focus a button -> tooltip appears -> move mouse over the tooltip or button -> tab to blur button.

should the mouse keep the tooltip alive until it moves out? 

**case 2:** focus a button -> tooltip appears -> move mouse over and then out.

should the tooltip disappear?

**Comment by StommePoes** [comment]
^ what do 80+% of users expect?

should they be trained to expect otherwise?

**Comment by idoros** [comment]
I can't say for 80% of users, but my intuition is that no matter how a tooltip opened, holding the mouse pointer on it is like a another focus point that even when the actual focus is moved a away, the content should persist.

Given this thinking, when the tooltip is invoked by the focus, a mouse might dismiss it by waving it away like `esc` would.

**Comment by carmacleod** [comment]
@idoros 
> what should be the cross behavior between focus and hover.

Good question. I don't know the answer. Your suggestions sound reasonable, but as @StommePoes said, you might want to do some user testing to see which behaviors users find the most intuitive. I'm not so sure about case 1 in a toolbar, though. If I shift+tab to the markdown toolbar for this github comment editor (H, **B**, _I_, etc) and use left/right arrow keys to move through the toolitems, I think I would prefer that the previous tooltip closes when I type arrow to focus the next toolitem (no matter where the mouse is), because I don't want to see two tooltips - I just want to see _the next one_. For case 2, if a user is using a mouse-controlled magnifier to read the previous tooltip and then they type tab, I think they might want the tooltip to close so that they can move on to the next focused element. I'm only guessing though.

**Comment by JAWS-test** [comment]
@idoros Perhaps the WCAG people have ideas on how it would be correct according to SC 1.4.13 (you could ask at https://github.com/w3c/wcag/issues)

**Comment by idoros** [comment]
@carmacleod, that's all good points, I'm trying to understand what would be a good default to start with. 

I wonder if there should be different behaviors:
1. under different contexts - keep overlay under certain UI, and not under others (I don't really like this concept)
2. according to content - keep overlay for complex content that a user might in the middle of reading
3. also should this be a developer configuration or somehow a user choice

I think that for now I would go with consistency that is easier to explain rather than a more complex behavior. Anyhow because there are 2 different input methods used together, then I assume the user has control over them and has made the choice to move the mouse over the tooltip (it's not that the tooltip just opened under the mouse), so the user can move the mouse away with to gain a predictable outcome. 

@JAWS-test - Thank you for the reference 🙏  (I have over 40 tabs with various specs and discussions on the subject, including the SC 1.4.13 somewhere), I will do some more reading and then see if there is a place to ask there as well.

**Comment by carmacleod** [comment]
> consistency that is easier to explain

Probably the most consistent and easiest to explain would be:
- if focused and keyboard is used to blur, close tooltip
- if hovered and mouse is moved away, close tooltip

**Comment by a11ydoer** [comment]
@mcking65 @zcorpan Do you know why we did not have this example in the APG?

**Comment by DavidMacDonald** [comment]
Agree it would be good to come to agreement and implement in APG, unless we see a blocker.

**Comment by ZoeBijl** [comment]
If anyone wants to collaborate on this, let me know via Twitter/email ✨. Would love to knock this out of the park in a zoom call or something ☺️.

**Comment by DavidMacDonald** [comment]
Hi @ZoeBijl 
When you say collaborate, are you thinking there is more to do on the example here 
https://s.codepen.io/Moiety/debug/LaPvWy
Seems OK to me... 
perhaps you see something I'm missing.
I think we might be able to drop in as is, and wait for comments :)

**Comment by jnurthen** [comment]
@DavidMacDonald one issue is that the tooltip always appears at the bottom of the icon whether there is space there or not. Most tooltip implementations will render the tooltip either above or below depending on the space currently available within the viewport.

**Comment by a11ydoer** [comment]
@DavidMacDonald Here is the [most recent tooltip discussion note](https://github.com/w3c/aria-practices/issues/128#issuecomment-900533821).

**Comment by ZoeBijl** [comment]
@DavidMacDonald the thing that _was_ missing was bounding box detection. When the control was too close to the window’s edge the tooltip would overflow outside of the window; making the text unreadable. @thezacharytaylor and I updated the component this week—it now does left and right bounding box detection and correction! We’ll see if we can do the same for top and bottom.

Should be ready fairly soon!

[Updated component](https://cdpn.io/pen/debug/LaPvWy) is at the same URL.

**Comment by ZoeBijl** [comment]
> I noticed that the tooltips when viewed on iPhone and open when touched, one tooltip will close when the other is opened. However, the tool tips do not close when touched again (opened) or when touching on the surrounding background. From a usability perspective how is a touch user meant to close an open tooltip?

Good point! I’m not entirely sure how to fix it. Maybe I can check for taps outside of the button/tooltip area?

**Comment by ZoeBijl** [comment]
Development is done. The [tooltip code is up on my GitHub](https://github.com/ZoeBijl/apg-tooltip). Feedback about their functionality and code clarity would be much appreciated! I’ll move it to a APG PR for further review after initial comments :)

**Comment by DavidMacDonald** [comment]
Awesome!
@ZoeBijl

**Comment by atoedt** [comment]
Dear ZoeBijl, 
thank you very much for this. We will try and implement that pattern for our Learning Management System abnd get back with feedabck as soon as we have it. You are my hero!
Alexandra

**Comment by SukumarSV** [comment]
Pressing` shift+F10 `as in G-docs and MS word opens the suggestion tools. 

**Tooltip:** On focus we can enable the tooltip with delay, and if the user press `shift+F10`the pop-up will become sticky until  the use `ESC` from it. Pressing down arrow will take the user inside the tool tip. As usual other controls will take place. 

If `aria-orientation` requires to direct the popup direction we use that as well.

**Comment by clshortfuse** [comment]
We are rewriting our tooltips with better focus on accessibility. We have a CSS only one that is similar to the @ZoeBijl example that includes a wrapper and target (eg: `.tooltip-wrapper` and `button.tooltip-target`). It uses hover and if primary pointer is coarse (touch), there's a 1.5 second delay (as recommended by Google in their Material Design framework). Keyboard users see it immediately if they tab into an textless (icon) button.

But I'd like to share some concerns we've have with the approach since we are in the process of rebuilding.

The system doesn't work in layouts that don't have `overlay:visible`. Tooltips can't float over scrollbars, and if you're not careful, will appear clipped. Touch detection also has to be Javascript, because there's not enough specificity in CSS for hybrid pointer environments (eg: touchscreen PCs).

So now, we've considered *tooltip* should be a top-level popup (ideally `HTMLPopupElement`), like how we've coded combobox and menus work (include self alignment). Then, it doesn't matter what spawned it.  There's also other minutia, such as, `mousemove` vs `mouseover`. Moving your cursor over a "tooltipable" element shouldn't spawn until it's still. (ie: `mousemove` resets counter). That would like all be folded into the same event listeners.

----

The other point is Escape to close. We do not implement this because it would be confusing for users who are focused on an icon-button within a form within a dialog. Those that could just be considered edge-cases though.

**Comment by JAWS-test** [comment]
@elizabethsjudd In my opinion, there is no role for it. The least confusing would be button. A button would also be appropriate if the tooltip is not opened automatically when the focus is received, but only after the button is activated. 

Or tooltips are only used on elements that already have a widget role. 

Alternatively, the results of the scan can simply be ignored, because the scans are not always right either. There is no WCAG rule that says that an element with tabindex must have a widget role.

**Comment by phoenisx** [comment]
Hi Folks!

Are tooltips not supposed to be clickable? I was working with some component libraries and (all of them which I worked with) do not make tooltip visible on click.
This is the same behaviour mentioned in https://www.w3.org/WAI/ARIA/apg/patterns/tooltip/ as well.

But if that's the case, how will someone be able to open a tooltip on Mobile devices, with touch only interactions?

**Comment by clshortfuse** [comment]
@phoenisx They shouldn't be interactive. Interactive one would be "toggletips" which are something else more aligned with popups. I know Android/Material Design shows tooltips after holding a button down for a short period of time.

**Comment by aleventhal** [comment]
Would be great to an example of rich tooltips using ARIA so that we can ensure they work with Chrome

**Comment by aleventhal** [comment]
Actually, any current page that works like a tooltip, even if it's not an official example. @mcking65 @keithamus anything on Facebook or GitHub that we can use?

**Comment by clshortfuse** [comment]
@aleventhal I have this setup, though not sure why it comes up as `aria-hidden`

![Image](https://github.com/user-attachments/assets/4080e95f-ed1c-4327-a69c-558fcecd39b1)

https://clshortfuse.github.io/materialdesignweb/components/buttons.html

**Comment by kpjfairns** [comment]
Can screenreaders automatically prioritise the tooltip, so that the blind can be guided on where these tooltips even are?

They may not even be able to read the tooltip unless integration with screenreaders is prioritised to some extent

This would probably mean a need to develop more from the screenreader perspective, being able to inject a tooltip as a secondary audio process will allow the text being read to continue, while also allowing the exploration of the webpage through the tab index, which should allow for tooltip enabled elements to be in focus independently of the rest of the text

It's the same problem for alerts and popup|model overlays

**Comment by keithamus** [comment]
A tooltip probably wants to use `popover=hint`, which will render on the top layer and respond to `Escape` keys. With the proposed [`interestfor`](https://github.com/whatwg/html/pull/11006) mapping this would create the hover interaction and automatic aria-details/aria-describedby associations (`aria-details` if the popover had any interactive children, otherwise `describedby`).
